sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -118,6 +118,31 @@ trait DateTimeFormatterHelper {
         s"before Spark 3.0, or set to CORRECTED and treat it as an invalid datetime string.", e)
   }
 
+  // When legacy time parser policy set to EXCEPTION, check whether we will get different results
+  // between legacy formatter and new formatter. If new formatter fails but legacy formatter works,
+  // throw a SparkUpgradeException. On the contrary, if the legacy policy set to CORRECTED,
+  // DateTimeParseException will address by the caller side.
+  protected def checkDiffFormatResult[T <: Date](
+      d: T,
+      legacyFormatFunc: T => String): PartialFunction[Throwable, String] = {
+    case e if needConvertToSparkUpgradeException(e) =>
+      val resultCandidate = try {
+      legacyFormatFunc(d)
+    } catch {
+      case _: Throwable => throw e
+    }
+    throw new SparkUpgradeException("3.0", s"Fail to format '$resultCandidate' in the new" +
+      s" formatter. You can set ${SQLConf.LEGACY_TIME_PARSER_POLICY.key} to LEGACY to restore" +
+      s" the behavior before Spark 3.0, or set to CORRECTED and treat it as an invalid" +
+      s" datetime string.", e)
+  }
+
+  private def needConvertToSparkUpgradeException(e: Throwable): Boolean = e match {
+    case _: DateTimeException | _: ArrayIndexOutOfBoundsException
```

Comment:
Shall we use Nonfatal here?

Reply:
This seems like a Java bug to me to throw `ArrayIndexOutOfBoundsException`.

Shall we fail the formatter creation when there are more than 10 `y` in the pattern string? And document this limitation as well.

Reply:
okay. Do we still need to check the runtime exceptions for the formatting phase?

Reply:
Yes, we still need to handle `DateTimeException` as the `DatetimeFormatter.format` method declares to throw `DateTimeException`, although I don't know how.